Add option for marking a user as spammer in "Deactivate user" popover

whilstsomebody opened the pull request:
This PR changes the workflow for marking user as a spammer.

## Tasks completed

- [x] At the bottom of the popover, a line that says: "Is this user a spammer?", with a checkbox (unchecked by default).
- [x] When the checkbox is checked, the popver expands with the following text:

Upon deactivation:
   
   * This user will be renamed to Deleted user.
   * The user's profile picture will be removed. 
 
- [x] A line that says: "What should be done with {user}'s messages?", with a dropdown. Options: Nothing, Delete messages in public channel, Delete all messages. If one of the "Delete messages" options is selected, the user's messages should be deleted as specified when the user is deactivated.

## Screenshots

![image](https://github.com/user-attachments/assets/cf23491f-99d8-42fa-bade-4ed4c56a6d44)

![image](https://github.com/user-attachments/assets/7efc8698-b51c-46be-8a27-e2cc0b14104a)

![image](https://github.com/user-attachments/assets/69cfdd9d-272b-48c9-8864-221ae06e39b7)

**User name is changed to "Deleted user".**

![image](https://github.com/user-attachments/assets/a4da011b-d919-44ef-aa46-6e89c4d52cc2)

![image](https://github.com/user-attachments/assets/bfce8ce4-a7f7-436e-acbc-1e40eff0201a)

**User's profile pricture is removed.**

![image](https://github.com/user-attachments/assets/ae8c194d-ba9f-42ca-afc1-fc44e2a949a0)

![image](https://github.com/user-attachments/assets/25f189e7-fdcb-42d4-bf3d-057223ed3c0f)


Fixes: #19841

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

whilstsomebody commented:
I will do further work on this one after 3 march due to exams.

whilstsomebody commented:
@zulipbot remove "has conflicts"

whilstsomebody commented:
@alya , could you review this one now?

whilstsomebody commented:
@alya , could you take a look at this?

alya commented:
Let's tweak the question about messages:

> Should messages sent by this user be deleted?
Don't delete <- default; Delete messages in public channels; Delete all messages

alya commented on web/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -40,3 +40,34 @@
+    <ul class="spammer-field-body">
+        <li>
+            {{#tr}}
+            This user will be renamed to <strong>Deleted user</strong>.
```
```suggestion
            They be renamed to <strong>Deleted user</strong>.
```

alya commented on web/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -40,3 +40,34 @@
+            This user will be renamed to <strong>Deleted user</strong>.
+            {{/tr}}
+        </li>
+        <li>{{t "The user's profile picture will be removed."}}</li>
```
```suggestion
        <li>{{t "Their profile picture will be removed."}}</li>
```

alya commented:
What are we doing with internationalization? Probably ideally they'd be renamed to "Deleted user" in the language of the organization.

Regardless, the string saying what they'll be renamed to needs to be accurate.

alya commented:
Seems reasonable in manual testing otherwise. I haven't tried to carefully validate the functionality.

alya commented:
@sahil839 Could you please take a look at this one?

sahil839 commented on zerver/actions/users.py:
```diff
@@ -512,7 +519,13 @@ def do_deactivate_user(
+                _cascade=False,
+                acting_user=acting_user,
+                is_spammer=is_spammer,
+                message_delete_action=message_delete_action,
```
This seems fine to consider bot as spammer if their owner is spammer and take same action on messages, but should we also do the things like renaming them or removing their profile picture, etc? Is there some discussion on how to handle bots owned by spammer? If not would be good to discuss it once.

sahil839 reviewed (commented):
Posted a batch of comments.

sahil839 commented on zerver/views/users.py:
```diff
@@ -119,6 +119,8 @@ def deactivate_user_backend(
     *,
     user_id: PathOnly[int],
+    is_spammer: Json[bool] = True,
+    message_delete_action: Json[int] = 0,
```
This should also use the constant and the default value for `is_spammer` should be `False`.

sahil839 replied:
Are we allowing deleting messages without marking the user as spammer? Here it is allowed but `do_deactivate_user` will not delete messages if `is_spammer` is not `True`.

sahil839 replied:
We can raise error if we do not want to allow. Can be discussed once on CZO.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -13466,6 +13466,31 @@ paths:
+            - No delete action: 0
+            - Delete messages in public stream: 1
+            - Delete all messages: 2
+              **Changes**: New in Zulip 10.0 (feature level 379).
```
There should be a blank line before "Changes" entry here as well.

sahil839 replied:
And it should be `Zulip 11.0`.

sahil839 commented on zerver/actions/users.py:
```diff
@@ -498,7 +500,12 @@ def send_events_for_user_deactivation(user_profile: UserProfile) -> None:
+    *,
+    acting_user: UserProfile | None,
+    is_spammer: bool = False,
+    message_delete_action: int = 0,
```
This should also use constant instead of `0`.

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -103,3 +106,31 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+    public_stream_messages = []
+
+    for message in messages:
+        if message.is_public_stream_message():
```
It is not efficient to do a DB query in loop here as there can be many messages. There should be a way to do bulk query using filter or something similar.

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -103,3 +106,31 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+    )
+    private_messages, public_stream_messages = classify_messages(user_profile, user_messages)
+    if message_delete_action == Message.DELETE_PUBLIC_STREAM_MESSAGE:
+        do_delete_messages(user_profile.realm, public_stream_messages, acting_user=user_profile)
```
Yes, as event data will not be correct, like `message_ids` field will contain messages not present in the `topic` field.

sahil839 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,12 @@ format used by the Zulip server that they are interacting with.
+
+* [`DELETE /users/{user_id}`](/api/deactivate-user): Add new
+  parameters `is_spammer` boolean and `message_delete_action` integer
+  for marking as spam while deactivating a user.
```
This isn't changed yet.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -1171,7 +1171,21 @@ export function show_edit_user_info_modal(user_id: number, $container: JQuery):
+
+            const NO_DELETE_ACTION = 0;
+            const DELETE_PUBLIC_STREAM_MESSAGES = 1;
+            const DELETE_ALL_MESSAGES = 2;
```
I think this can be defined probably in `settings_config`. These are not related to settings, but we have other constants like these defined there. And then these constants can be directly used as value for the `option` elements instead of using strings like "delete-public", etc.

sahil839 commented on web/src/user_deactivation_ui.ts:
```diff
@@ -79,6 +79,26 @@ export function confirm_deactivation(
+            function set_spammer_field_visibility(dialog_widget_id: string): void {
+                const $modal = $(`#${CSS.escape(dialog_widget_id)}`);
+                const $spammer_checkbox = $modal.find(".spammer");
+                const $spammer_field = $modal.find(".spammer-actions"); // Updated class name
```
This variable can also be named as `$spammer_actions_container`.

I don't get the comment here.

sahil839 commented on web/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -42,3 +42,34 @@
+        <label for="message-action">
+            {{#tr}}
+                Should messages sent by this user be deleted?
+            {{/tr}}
```
This can use `{{t "Should messages ..."}}`.

sahil839 reviewed (commented):
I posted some comments, but I guess this would move forward once the preparatory PR for updating message deletion code will be completed.

sahil839 commented on zerver/actions/users.py:
```diff
@@ -528,6 +540,31 @@ def do_deactivate_user(
+                    full_name="Deactivated bot"
+                )
+
+            delete_deactivated_user_messages(user_profile, message_delete_action)
```
But it is possible when making an API request using a tool like curl.

sahil839 replied:
I think the above linked PR is not related to this. And still messages are only deleted when marking user as spammer and I guess it was discussed on CZO to allow deleting messages without marking them as spammer.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -1171,7 +1171,24 @@ export function show_edit_user_info_modal(user_id: number, $container: JQuery):
+            } else if (
+                $("#message-delete-action").val() ===
+                message_delete_actions.delete_all_messages.value
+            ) {
```
That was different from what I am suggesting now. I am telling the `<option>` element to have value set to "0". "1", etc. and then get the value here like `Number.parseInt($("#message-delete-action").val(), 10)`.

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -103,3 +107,35 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+
+    for message in messages:
+        if message.recipient.type == Recipient.STREAM:
+            stream = Stream.objects.get(id=message.recipient.type_id)
```
I am not sure if I mentioned this before, but this will result in DB queries in a loop which is not efficient.

sahil839 replied:
We do not need whole stream objects, so would be good if we could just fetch fields needed like or maybe we can just fetch IDs of all public streams and then check whether the stream ID is in that set or not.

sahil839 replied:
And there is still a query in a loop for fetching `Stream` object which will be not needed after the above change.

sahil839 reviewed (commented):
Posted a few comments, but I think it would be good to review this completely after #34545 is merged.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -14104,6 +14104,32 @@ paths:
+            - Delete messages in public stream: 1
+            - Delete all messages: 2
+
+            **Changes**: New in Zulip 11.0 (feature level {ZF-01a225}).
```
Braces are not required around the "ZF-...".

sahil839 commented on api_docs/unmerged.d/ZF-01a225:
```diff
@@ -0,0 +1,3 @@
+* [`DELETE /users/{user_id}`](/api/deactivate-user): Add new
```
This should be "Added new ..."

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -103,3 +106,31 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+        do_delete_messages(user_profile.realm, public_stream_messages, acting_user=user_profile)
+    elif message_delete_action == Message.DELETE_ALL_MESSAGE:
+        do_delete_messages(user_profile.realm, private_messages, acting_user=user_profile)
+        do_delete_messages(user_profile.realm, public_stream_messages, acting_user=user_profile)
```
How are DMs being handled now? Comment in `do_delete_messages` mention that DMs should be grouped. Also, there is no need to classify public and private stream messages for this case I guess.

whilstsomebody commented:
Hey @sahil839 , I have addressed all the review comments. Following is the summary of the changes:

- Changes the name of `classify_messages` function to `get_public_stream_messages` because now we are not concerned about grouping the public and private messages as the `do_delete_messages` function now groups the messages. 
`get_public_stream_messages` now just returns the private stream messages.

- Now we only fetch the stream messages in `delete_deactivated_user_messages` and pass the whole list if case of the third option, else use the `get_public_stream_messages` function for the second option.

Please take a look. Thanks!

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -146,3 +147,44 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
         move_messages_to_archive(message_ids, chunk_size=retention.STREAM_MESSAGE_BATCH_SIZE)
+
+
+def get_public_stream_messages(realm: Realm, messages: Iterable[Message]) -> list[Message]:
```
I think this function should be in `zerver/lib/message.py` and maybe rename this something like `filter_public_stream_messages`  because we are passing messages to the function.

sahil839 commented on web/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -42,3 +42,32 @@
+        </label>
+        <div class="input-group">
+            <select id="message-delete-action" class="spammer-message-action-dropdown modal_select bootstrap-style-font">
+                <option value="0">{{t "Don't delete" }}</option>
```
We should define these 0,1 and 2 values in `settings_config.ts` like an object named something like `message_delete_action` which can have keys like `delete_none`, `delete_all_public_stream_messages`, etc.

whilstsomebody replied:
But in https://github.com/zulip/zulip/pull/33561#discussion_r2125853513, you said to get the values directly from the select elemts.

sahil839 replied:
I meant there is no need to use special string like "no_delete", etc. Value should be 0 but it should be defined to a constant.

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -146,3 +147,44 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+        public_stream_messages = get_public_stream_messages(realm, user_stream_messages)
+        do_delete_messages(user_profile.realm, public_stream_messages, acting_user=user_profile)
+    elif message_delete_action == Message.DELETE_ALL_MESSAGES:
+        do_delete_messages(user_profile.realm, user_stream_messages, acting_user=user_profile)
```
Was this discussed to not delete DMs? If yes, then I guess this should me mentioned in the docs clearly.

whilstsomebody replied:
Asked here: [#design > Add option to mark user as a spammer. @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/Add.20option.20to.20mark.20user.20as.20a.20spammer.2E/near/2203091)

sahil839 commented on zerver/views/users.py:
```diff
@@ -162,9 +166,16 @@ def _deactivate_user_profile_backend(
     target: UserProfile,
     *,
+    is_spammer: Json[bool] = False,
+    message_delete_action: Json[list[int]] | None = None,
```
I am not sure of this field name now that we pass a list of integers. Would be good to discuss this on #api-design.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -14173,6 +14173,37 @@ paths:
+
+            - No delete action: 0
+            - Delete messages in public streams: 1
+            - Delete messages in public streams: 2
```
I guess there is some typo here, both 1 and 2 mention public streams.

whilstsomebody commented:
Hey @sahil839 , I have made the required changes.

I have converted `message_delete_action` ,which was earlier a `list[int]`, to an object of `key:boolean` (`MessageDeleteAction`) as suggested here: [#api design > Marking a user as spammer @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/Marking.20a.20user.20as.20spammer/near/2210255)

I have question : in `zerver/views/_deactivate_user_profile_backend`, I am accepting `message_delete_action` as an object but then converting it to `list[int]` only before passiing to `zerver/actions/do_deactivate_user`. Is this fine?

Could you please take a look?
Thanks!

sahil839 commented on zerver/tests/test_users.py:
```diff
@@ -1950,6 +1951,197 @@ def test_email_sent(self) -> None:
+            user, private_stream.name, topic_name="topic2", content="private stream message 2"
+        )
+
+    def test_api_with_no_delete_action(self) -> None:
```
Test functions for message deletion should mention about deactivation to make it more clear like `test_deactivation_without_deleting_messages`, `test_deactivation_with_deleting_public_stream_messages`, etc.

sahil839 commented on zerver/views/users.py:
```diff
@@ -162,9 +172,30 @@ def _deactivate_user_profile_backend(
+
+        # If no actions selected, default to no delete
+        if not message_delete_action_list:
+            message_delete_action_list = [Message.NO_DELETE_ACTION]
```
I think it would be simpler to just pass the object as it is since all three cases are being handled separately. For case of no delete action, `delete_deactivated_user_messages` can just return early if all the fields are false.

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -146,3 +147,33 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+                recipient__type__in=[Recipient.PERSONAL, Recipient.DIRECT_MESSAGE_GROUP],
+            )
+        )
+        do_delete_messages(user_profile.realm, user_direct_messages, acting_user=user_profile)
```
Comment in `do_delete_messages` mentions that caller will need to handle 1:1 DMs. And reading the code it seems that you would need to handled 1:1 DM with each user separately, otherwise users other than the sender and receiver will get the events. You should check this once if that is the case and then you would need to update that accordingly.

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -146,3 +147,33 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+    if Message.DELETE_PUBLIC_STREAM_MESSAGES in message_delete_action:
+        do_delete_messages(user_profile.realm, public_stream_messages, acting_user=user_profile)
+    if Message.DELETE_PRIVATE_STREAM_MESSAGES in message_delete_action:
+        do_delete_messages(user_profile.realm, private_stream_messages, acting_user=user_profile)
```
I guess `do_delete_messages` can be called just once atleast for private and public stream messages since `do_delete_messages` handles the groupings. Maybe you can part streams and group DMs together to `do_delete_messages` and 1:1 DMs separately by grouping as per the other user in 1:1 DM.

sahil839 replied:
Now `do_delete_messages` fetches each stream in a different query when messages from multiple streams are passed together, so maybe we can pass it an optional parameter in the form of a `dict[int, Stream]` so that it can be used directly.

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -146,3 +147,33 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+) -> None:
+    if Message.NO_DELETE_ACTION in message_delete_action:
+        return
+    user_stream_messages = list(
```
Maybe we can avoid this query if only direct messages are to be deleted. Or just fetch all the messages once together and then classify it based on what messages user wants to delete.

sahil839 replied:
`do_delete_messages` calls `is_stream_messages` now which uses `self.recipient.type` so it would be good to prefetch `recipient` field using `select_related` in `Message` query.

whilstsomebody commented:
Hey @sahil839 , I have made the changes you suggested. Below is the summary:

- Renamed the test functions.
- Now we don't use `list[int]` anywhere. 
- Removed : 
```py
NO_DELETE_ACTION,
DELETE_PUBLIC_STREAM_MESSAGES,
DELETE_PRIVATE_STREAM_MESSAGES,
DELETE_DIRECT_MESSAGES,
```
- Added an optional paramater `stream_by_recipient_id` to `do_delete_messages` as suggested here : https://github.com/zulip/zulip/pull/33561#discussion_r2185566835
- Created a new function `_group_personal_messages_by_conversation` to group 1:1 DMs (not sure if I have done this correctly or not)
- https://github.com/zulip/zulip/pull/33561#discussion_r2185564121 : Not getting all the messages at once but yes, calling it separately for stream messages and DMs.

Could you take a look at the changes? Thanks!

whilstsomebody commented:
Hey @sahil839 , I have made the changes you suggested. Below is the summary:

- Renamed the test functions.
- Now we don't use `list[int]` anywhere. 
- Removed : 
```py
NO_DELETE_ACTION,
DELETE_PUBLIC_STREAM_MESSAGES,
DELETE_PRIVATE_STREAM_MESSAGES,
DELETE_DIRECT_MESSAGES,
```
- Added an optional paramater `stream_by_recipient_id` to `do_delete_messages` as suggested here : https://github.com/zulip/zulip/pull/33561#discussion_r2185566835
- Created a new function `_group_personal_messages_by_conversation` to group 1:1 DMs (not sure if I have done this correctly or not)
- https://github.com/zulip/zulip/pull/33561#discussion_r2185564121 : Not getting all the messages at once but yes, calling it separately for stream messages and DMs.
- Added some more tests to test the grouping logic.

Could you take a look at the changes? Thanks!

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -153,3 +160,97 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+
+
+def _group_personal_messages_by_conversation(
+    user_direct_messages: list[Message], sender_id: int
```
`sender_id` parameter is not used here.

sahil839 commented on zerver/lib/message.py:
```diff
@@ -1745,3 +1745,25 @@ def is_1_to_1_message(message: Message) -> bool:
+            realm=realm,
+            invite_only=False,
+            history_public_to_subscribers=True,
+            deactivated=False,
```
I think we can delete messages from deactivated streams as well.

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -96,20 +96,27 @@ def _process_grouped_messages_deletion(
+    messages: Iterable[Message],
+    *,
+    acting_user: UserProfile | None,
+    streams_by_recipient_id: dict[int, Stream] | None = None,
```
I think default for this can be an empty object instead of `None` and then `stream_by_recipient_id = streams_by_recipient_id or {}` line can be removed.

whilstsomebody replied:
For this change, linter throws error instructing that we should not use "mutable data structures".
Can we do something else?

sahil839 replied:
Ok. We could use mapping but this seems fine for now.

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -153,3 +160,97 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+                messages_to_delete.extend(private_stream_messages)
+
+        if messages_to_delete:
+            do_delete_messages(user_profile.realm, messages_to_delete, acting_user=user_profile)
```
I think `acting_user` here should be set as user deactivating the user not the user being deactivated.

sahil839 commented on zerver/actions/message_delete.py:
```diff
@@ -153,3 +160,97 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+            if message_delete_action.get("delete_private_stream_messages", False):
+                messages_to_delete.extend(private_stream_messages)
+
+        if messages_to_delete:
```
I have refactored this code somewhat which feels much more readable to me including some changes suggested. There might be some syntax error or variable name bugs, but would love to know your thoughts on this.
```py
def delete_deactivated_user_messages(
    realm: Realm, user_profile: UserProfile, message_delete_action: MessageDeleteAction, acting_user: UserProfile
) -> None:
    delete_public_stream_messages = message_delete_action.get("delete_public_stream_messages", False)
    delete_private_stream_messages = message_delete_action.get("delete_private_stream_messages", False)
    delete_direct_messages = message_delete_action.get("delete_direct_messages", False)

    if not (
        delete_public_stream_messages
        or delete_private_stream_messages
        or delete_direct_messages
    ):
        return

    message_filter_query = Q()
    message_exclude_query = Q()

    if delete_direct_messages:
        # 1:1 DMs will be handled separately.
        message_filter_query |= Q(recipient__type=Recipient.DIRECT_MESSAGE_GROUP)

    if delete_public_stream_messages and delete_private_stream_messages:
        message_filter_query |= Q(recipient__type=Recipient.STREAM)
    elif delete_public_stream_messages or delete_private_stream_messages:
        public_stream_ids = get_public_streams_queryset(realm).values_list("id", flat=True)
        if delete_public_stream_messages:
            message_filter_query |= Q(recipient__type=Recipient.STREAM, recipient__type_id__in=public_stream_ids)
        else:
            message_filter_query |= Q(recipient__type=Recipient.STREAM)
            message_exclude_query |= Q(recipient__type=Recipient.STREAM, recipient__type_id__in=public_stream_ids)

    messages = Message.objects.filter(sender=user_profile).filter(message_filter_query).exclude(message_exclude_query)
    do_delete_messages(user_profile.realm, messages, acting_user=acting_user)
    
    # 1:1 DMs need to be handled separately as we need to group them to
    # call do_delete_messages for each conversation separately.
    if delete_direct_messages:
        direct_messages = Message.objects.filter(sender=user_profile, recipient__type=Recipient.PERSONAL)

        personal_messages_by_recipient_dict: defaultdict[int, list[Message]] = defaultdict(list)
        for message in direct_messages:
            recipient_user_id = message.recipient.type_id
            personal_messages_by_recipient_dict[recipient_user_id].append(message)

        for conversation_messages in personal_messages_by_recipient_dict.values():
            do_delete_messages(user_profile.realm, conversation_messages, acting_user=acting_user)
```

whilstsomebody replied:
One more thing, I think the `delete_direct_messages` block should be like this:

```py
if delete_direct_messages:
        direct_messages = list(
            Message.objects.filter(
                sender=user_profile, 
                realm_id=user_profile.realm_id,
                recipient__type=Recipient.PERSONAL
            ).select_related("recipient")
        )

        personal_messages_by_other_user, direct_message_group_messages = _group_personal_messages_by_conversation(
            direct_messages
        )

        # Process 1:1 DMs grouped by conversation
        for conversation_messages in personal_messages_by_other_user.values():
            do_delete_messages(user_profile.realm, conversation_messages, acting_user=acting_user)

        # Process any group DMs that were mixed in with the direct messages query
        if direct_message_group_messages:
            do_delete_messages(user_profile.realm, direct_message_group_messages, acting_user=acting_user)

```

sahil839 replied:
I think we do not need a check for message_filter_query since we already define it at start as `Q()`.

And for the DMs, I guess we can handle group DMs together with stream messages since `do_delete_messages` is capable of handling them together and would save us in having a comparitively complex function for grouping group DMs and 1:1 DMs.

whilstsomebody commented on zerver/views/users.py:
```diff
@@ -109,6 +110,12 @@
+class MessageDeleteAction(BaseModel):
+    delete_public_stream_messages: bool = False
+    delete_private_stream_messages: bool = False
+    delete_direct_messages: bool = False
```
I am facing issue with this. According to me, both of them are required since they serve different purposes in the application architecture. Could you explain how to approach here?

sahil839 replied:
Pydantic can validate TypedDict just like it validates `dict` objects. We use them in `views/video_calls.py`, just that we need to import TypedDict from `typing_extensions`. And we also use `dataclass` for `GroupSettingChangeRequest` which works fine with pydantic too.

whilstsomebody commented:
Hey @sahil839 , I have addressed you reviews and have just one [question](https://github.com/zulip/zulip/pull/33561#discussion_r2195483184). Please take a look.

whilstsomebody commented:
Current Status: 

- Removed the `MessageDeleteAction` Basemodel class that was defined in `/views/users.py`;instead using `MessageDeleteAction` `Typedict` class defined in `/actions/message_delete.py`.
- Removed the `classify_stream_messages` that was defined in `/lib/message.py` since we are not using it anymore.
- Updated `delete_deactivated_user_messages` function as suggested above.

@sahil839, could you take a look at this again? Thanks!

sahil839 commented on web/src/user_deactivation_ui.ts:
```diff
@@ -61,6 +67,7 @@ export function confirm_deactivation(
                 admin_email: people.my_current_email(),
                 realm_url,
                 realm_name,
+                message_delete_actions: get_message_delete_actions(),
```
I guess this can directly call `Object.values(settings_config.message_delete_action_values)`. Or better to just have `settings_config.message_delete_action_values` as a list instead of an object.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -1168,7 +1168,22 @@ export function show_edit_user_info_modal(user_id: number, $container: JQuery):
-            dialog_widget.submit_api_request(channel.del, url, {});
+            const is_spammer = $("#is-user-spammer").is(":checked");
+            const message_delete_action = {
+                delete_public_stream_messages: $("#message-delete-action[value='1']").is(
```
We should avoid using hardcoded values like 1,2 here. You can use a loop over `settings_config.message_delete_action_values` and use `key` and `code` fields to create the `message_delete_action` object.

whilstsomebody commented:
@sahil839 , I have addressed the above reviews. Please take a look.

sahil839 commented on web/src/settings_config.ts:
```diff
@@ -318,11 +318,14 @@ export const get_realm_topics_policy_values = (): RealmTopicsPolicyValues => {
+            description: $t(
+                {defaultMessage: '"{empty_topic_name}" topic allowed'},
+                {empty_topic_name},
+            ),
```
These are some unrelated changes.

sahil839 reviewed (commented):
Posted one comment. Also, user is not being marked as spammer when the checkbox is checked when deactivating directly using the icon button in User list row.

whilstsomebody commented:
Hey @sahil839 , I have fixed the above mentioned bug. Could you take a look? 
Thanks!

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -1168,13 +1168,24 @@ export function show_edit_user_info_modal(user_id: number, $container: JQuery):
 
     // Handle reactivation
-    $("#edit-user-form").on("click", ".reactivate_user_button", (e) => {
+    $("body").on("click", ".reactivate_user_button", (e) => {
```
Why is this changed? I guess we should have a prep commit for this with commit message explaining why this change is needed and why it is correct.

whilstsomebody replied:
We don't this change anymore, I have already fixed this in https://github.com/zulip/zulip/pull/35290.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -1168,13 +1168,24 @@ export function show_edit_user_info_modal(user_id: number, $container: JQuery):
+                is_spammer,
+                message_delete_action: JSON.stringify(message_delete_action),
+            };
+            dialog_widget.submit_api_request(channel.del, url, data);
```
I think this should ideally be a function defined just once and not twice. Might be good to do it as a follow-up.

whilstsomebody commented:
@sahil839 , I have addressed you review. Please take a look. Thanks.

sahil839 commented on api_docs/unmerged.d/ZF-01a225:
```diff
@@ -0,0 +1,3 @@
+* [`DELETE /users/{user_id}`](/api/deactivate-user): Added new
+  parameters `is_spammer` and `message_delete_action` for marking 
```
`message_delete_action` parameter is used to delete messages irrespective of whether user is being marked as spammer or not.

sahil839 reviewed (commented):
Posted one comment. Looks good otherwise.

sahil839 commented:
I fixed a bug in #35325 so you would need to rebase this PR. Looks good to me otherwise.

whilstsomebody commented:
@zulipbot remove "maintainer review"  and add "integration review"

whilstsomebody commented:
@zulipbot remove "maintainer review"
@zulipbot add "integration review"

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -14507,6 +14507,50 @@ paths:
+                type: boolean
+              example: true
+          required: false
+        - name: message_delete_action
```
Should we consider calling this just `action`, to provide the opportunity to extend later to actions that aren't specifically deleting messages?

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -14507,6 +14507,50 @@ paths:
+            An object with boolean values for each message type:
+
+            - `delete_public_stream_messages`: Delete messages in public streams
+            - `delete_private_stream_messages`: Delete messages in private streams
```
`Delete messages in private streams => Delete messages in private channels` to follow our policy that we use "channels" in all documentation strings to refer to them.

Also, probably `delete_private_stream_messages` should be `delete_private_channel_messages` to avoid having to rename that later as we move the API towards using "channels" for variable names too.

timabbott commented on zerver/views/users.py:
```diff
@@ -134,6 +137,8 @@ def deactivate_user_backend(
         user_profile,
         target,
+        is_spammer=is_spammer,
+        message_delete_action=message_delete_action,
```
Somewhere we need to be sure that the organization's `can_delete_message_group` policy is checked for the acting user.

timabbott commented on zerver/actions/message_delete.py:
```diff
@@ -153,3 +163,70 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+    delete_private_stream_messages = message_delete_action.get(
+        "delete_private_stream_messages", False
+    )
+    delete_direct_messages = message_delete_action.get("delete_direct_messages", False)
```
I think we wouldn't need this code if we defined `message_delete_action` as a Pydantic dataclass with default values of `false` for each of these fields.

Additionally, I think we could call `message_delete_action` something like `deactivate_user_actions: DeactivateUserActions`

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -1950,6 +1951,270 @@ def test_email_sent(self) -> None:
         self.assertEqual(msg.reply_to[0], "noreply@testserver")
         self.assertIn("Dear Hamlet,", msg.body)
 
+    # Helper for setting up various types of messages.
```
Could we make these tests a test class class, `DeactivateSpammerTest`?

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -1950,6 +1951,270 @@ def test_email_sent(self) -> None:
+
+    def test_deactivation_without_deleting_messages(self) -> None:
+        admin = self.example_user("othello")
+        do_change_user_role(admin, UserProfile.ROLE_REALM_ADMINISTRATOR, acting_user=None)
```
We should just use Iago, the user defined in our test database to always already be an administrator.

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -1950,6 +1951,270 @@ def test_email_sent(self) -> None:
+        )
+        self.assert_json_success(result)
+        user_messages = list(get_user_sent_messages(user))
+        self.assert_length(user_messages, 8)  # No messages should be deleted
```
Don't put comments on the end of a line; put them on the line before what they're describing.

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -1950,6 +1951,270 @@ def test_email_sent(self) -> None:
+        )
+        self.assert_json_success(result)
+        user_messages = list(get_user_sent_messages(user))
+        self.assert_length(user_messages, 4)  # 4 stream messages deleted out of 8 messages.
```
On a line by itself, this comment might read:

> 2 private channel messages and 2 public channel messages were deleted; the 4 DMs remain.

whilstsomebody commented on zerver/actions/users.py:
```diff
@@ -533,6 +545,24 @@ def do_deactivate_user(
+            elif user_profile.full_name != "Deactivated bot" and user_profile.is_bot:
+                do_change_full_name(user_profile, "Deactivated bot", acting_user=acting_user)
+
+            # TODO: Change avatar image to that of an inaccessible user.
```
Since the TODO will only be relevant after this PR is merged, I think we should do it later.

sahil839 commented on zerver/tests/test_users.py:
```diff
@@ -2283,6 +2287,339 @@ def test_deliver_scheduled_emails_no_addressees(self) -> None:
+            },
+        )
+        self.assert_json_success(result)
+        self.assertTrue(cordelia.DoesNotExist)
```
Not sure how I missed this before, but how does this checks whether the user is deactivated or not. `user.DoesNotExist` is just a class as it prints `<class 'zerver.models.users.UserProfile.DoesNotExist'>`.

Correct way is to check `is_active` field.

whilstsomebody replied:
Ohh, actually I assumed it does what I want because of it's name. i have updated to use `is_active` now.

whilstsomebody commented:
@sahil839 , I have addressed the above comments. 

In the template `confirm_deactivate_user.hbs`, I have removed the id. Using the value only.

sahil839 commented on web/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -42,3 +42,41 @@
+    </ul>
+    {{#if user_can_delete_any_message}}
+    <div class="delete-message-handling">
+        <label for="message-action">
```
I guess `for` attribute can be removed here since it is set to the ID of the element it is referring but here we do not have ID for the input elements and there are multiple input elements.

sahil839 commented on web/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -42,3 +42,41 @@
+        <div class="input-group">
+            {{#each message_delete_actions}}
+                <label class="checkbox">
+                    <input value="{{this.key}}" type="checkbox"/>
```
Setting `value` attribute to a string for a checkbox type input seems odd. I think would be better to use `key` as class and then use that class to reference the element in JS.

whilstsomebody commented:
@sahil839 , I have addressed the above comments.

sahil839 commented:
I did some minor updates in frontend code-
- `is_spammer` field value is passed as JSON encoded boolean.
- Removed some comments that are not needed.
- `settings_config.message_delete_action_values` uses `description` key instead of `value` for the checkbox label.

sahil839 commented:
@timabbott ready for you review.

alya commented:
The spacing between the question and the checkboxes seems too tight, at least at small font sizes, and uneven with the spacing *between* checkboxes.
<img width="654" height="216" alt="Screenshot 2025-09-05 at 15 55 42@2x" src="https://github.com/user-attachments/assets/4961cd17-95f3-48b9-a6aa-9b33d8c7bfd3" />

alya commented:
Not super important, but the placeholder doesn't seem to update properly:

<img width="2228" height="300" alt="Screenshot 2025-09-05 at 16 04 05@2x" src="https://github.com/user-attachments/assets/d0345b76-4873-45f6-b6c2-c0ca524c78f3" />

alya commented:
Anyway, those are tiny issues from manual testing, and this is ready for @timabbott to take a look from my perspective.

alya commented:
Hm, I'd expect the spacing we want to depend on font size.

sahil839 commented:
There was a live update bug when deleting the private messages of the deactivated user because message deletion events were sent to the acting_user for DMs that the user was not involved in and thus cannot access. I added a preparatory commit to fix that.

timabbott commented on zerver/actions/message_delete.py:
```diff
@@ -85,10 +85,6 @@ def _process_grouped_messages_deletion(
 
-    if acting_user is not None:
-        # Always send event to the user who deleted the message.
-        users_to_notify.add(acting_user.id)
```
I think removing this would introduce a bug. As noted in https://github.com/zulip/zulip/commit/7a80fcf042cb18d2d8dd86ffb82f0c7eff59ad61, the purpose of this block is that if you're deleting a message in the Zulip UI, say in a channel you're not subscribed to, you get the appropriate live-update event for the message being deleted. 

Even if that analysis is wrong, removing this would be an API change that needs to be discussed in "api design" and documented as such, since it's exactly reverting a previous discussed API change.

timabbott commented on zerver/actions/users.py:
```diff
@@ -485,13 +495,39 @@ def do_deactivate_user(
+            if user_profile.full_name != "Deleted user" and not user_profile.is_bot:
+                do_change_full_name(user_profile, "Deleted user", acting_user=acting_user)
+            elif user_profile.full_name != "Deactivated bot" and user_profile.is_bot:
+                do_change_full_name(user_profile, "Deactivated bot", acting_user=acting_user)
```
Did we discuss whether these fixed names should be internationalized? I suspect the answer is it'd be risky to change that in this PR, but I guess maybe we should discuss in #backend and consider filing an issue if it seems like a safe change to make.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -15090,6 +15090,59 @@ paths:
+
+            - `delete_public_stream_messages`: Delete messages in public streams
+            - `delete_private_stream_messages`: Delete messages in private streams
+            - `delete_direct_messages`: Delete direct messages
```
This part of the API documentation seems useless, since it's a duplicate of the `content` section below?

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -15090,6 +15090,59 @@ paths:
         given their user ID.
```
Probably it's worth adding a sentence of the form:

> If the user owns any bots, those bots will be deactivated before deactivating the user. Clients are recommended to ask the administrator if they would like to change the owner for the bots before deactivating them.
